### PR TITLE
add support for sending and receiving raw binary data using []byte channels

### PR DESCRIPTION
**Pull request by michael2008**
This is mainly for supporting customized serialization methods like Protobuf and reduce message size in performance-critical scenarios.

### CODE REVIEW

File: sockets.go [hunk 1]
```diff
@@ -157,6 +157,49 @@ type JSONConnection struct {
...
+//   error or a client disconnect.
+// - A sending done channel  (chan<- bool) on which you can send as soon as you wish
+//   to disconnect the connection.
+// The middleware handles the following for you:
+// - Checking the request for cross origin access
+// - Doing the websocket handshake
+// - Setting sensible options for the Gorilla websocket connection
+// - Starting and terminating the necessary goroutines
+// An optional sockets.Options object can be passed to Messages to overwrite
+// default options mentioned in the documentation of the Options object.
```
Comment (beatrichartz) [inline sockets.go]: This documentation needs updating so it reflects that this is about `[]byte` channels

File: README.md [hunk 2]
```diff
@@ -16,6 +16,10 @@ Package `sockets` makes it fun to use websockets with Martini. Its aim is to pro
 
 `sockets.Messages` is a simple middleware that organizes websockets messages into string channels.
 
+#### Binary Messages
+
+`sockets.BinaryMessages` is a simple middleware that organizes websockets messages into []byte channels. The lower-level Gorilla Websocket supports sending raw bytes, and the high-level Martini middleware should also support it. The binary message channel is quite useful, where you can directly send your data using customized serialization methods like Protobuf. Message size is crucial in some use cases.
+
```
Comment (beatrichartz) [inline README.md]: This does not need to argue its usefulness that much, i.e. you could remove:

> The lower-level Gorilla Websocket supports sending raw bytes, and the high-level Martini middleware should also support it. The binary message channel is quite useful, where you can directly send your data using customized serialization methods like Protobuf. Message size is crucial in some use cases.

File: sockets.go [hunk 3]
```diff
@@ -157,6 +157,49 @@ type JSONConnection struct {
 	Receiver reflect.Value
 }
 
+// Binary Connection connects a websocket message connection to a []byte
+// channel.
+type BinaryConnection struct {
```
Comment (beatrichartz) [inline sockets.go]: This should be renamed to e.g. `ByteSliceConnection`. `BinaryConnection` suggests a `[]bool`, but we have a `[]byte`

File: sockets.go [hunk 4]
```diff
@@ -157,6 +157,49 @@ type JSONConnection struct {
...
+	*Connection
+
+	// Sender is the []byte channel used for sending out binary data to the client.
+	// This channel gets mapped for the next handler to use and is asynchronous
+	// unless the SendChannelBuffer is set to 0.
+	Sender chan []byte
+
+	// Receiver is the []byte channel used for receiving binary data from the client.
```
Comment (beatrichartz) [inline sockets.go]: Similarly here this is not specific to receiving binary data. I know the motivation for this is protobufs, but technically this is agnostic to what the byte slice represents.

File: sockets.go [hunk 5]
```diff
@@ -157,6 +157,49 @@ type JSONConnection struct {
...
+// - A sending done channel  (chan<- bool) on which you can send as soon as you wish
+//   to disconnect the connection.
+// The middleware handles the following for you:
+// - Checking the request for cross origin access
+// - Doing the websocket handshake
+// - Setting sensible options for the Gorilla websocket connection
+// - Starting and terminating the necessary goroutines
+// An optional sockets.Options object can be passed to Messages to overwrite
+// default options mentioned in the documentation of the Options object.
+func BinaryMessages(options ...*Options) martini.Handler {
```
Comment (beatrichartz) [inline sockets.go]: As per the other comments, this should be something like `ByteSliceMessages`

File: sockets.go [hunk 6]
```diff
@@ -606,6 +748,8 @@ func waitForDisconnect(c Binding) {
 func newBinding(iFace interface{}, ws *websocket.Conn, o *Options) Binding {
 	typ := reflect.TypeOf(iFace)
 
+	//fmt.Sprintln(typ.Kind().String())
+
```
Comment (beatrichartz) [inline sockets.go]: This should be removed

File: sockets_test.go [hunk 7]
```diff
@@ -19,7 +20,9 @@ const (
 	sendPath        string = "/sender"
 	pingPath        string = "/ping"
 	recvStringsPath string = "/strings/receiver"
+	recvBinaryPath  string = "/binary/receiver"
```
Comment (beatrichartz) [inline sockets_test.go]: As per the other comments, this should be something like `"/byteslice/receiver"`

File: sockets_test.go [hunk 8]
```diff
@@ -19,7 +20,9 @@ const (
...
 	sendStringsPath string = "/strings/sender"
+	sendBinarysPath string = "/binary/sender"
```
Comment (beatrichartz) [inline sockets_test.go]: same here

File: sockets_test.go [hunk 9]
```diff
@@ -29,19 +32,25 @@ type Message struct {
...
+	recvBinary         [][]byte
+	recvStringsCount   int
+	recvBinaryCount    int
+	recvStringsDone    bool
+	recvBinrayDone     bool
+	sendStrings        []string
+	sendBytes          [][]byte
+	sendStringsCount   int
+	sendStringsDone    bool
+	sendBinaryDone     bool
```
Comment (beatrichartz) [inline sockets_test.go]: This section needs renaming for byteslices as well

File: sockets_test.go [hunk 10]
```diff
@@ -279,6 +371,33 @@ func TestJSONReceive(t *testing.T) {
...
+			ws.Close()
+			return
+		}
+	}
+
+	expectBinariesToHaveArrived(t, 3, recvBinary)
+	expectStatusCode(t, http.StatusSwitchingProtocols, resp.StatusCode)
+	expectIsDone(t, recvBinrayDone)
+}
+
```
Comment (beatrichartz) [inline sockets_test.go]: same for the tests.

**Review by beatrichartz — CHANGES_REQUESTED** [review]
Thanks for this, looking good overall - nice testing, and a valuable addition to this library. There are some minor consistency issues, and the commits should be squashed.

I think the main thing to change is how we name the new interface. I know the motivation for this comes from protobufs, so you are thinking about binary messages. However, byte slices are not restricted to a particular serialization format, so I think we should rename this to something like `sockets.ByteSliceMessage` or similar. What are your thoughts?

**Comment by michael2008** [comment]
ByteSliceConnection is good and suitable for golang naming convention. Also, I will revise the docs to reflect these changes.

**Review by beatrichartz — APPROVED** [review]

**Pull request merged**